retention_rule.go

```diff
@@ -0,0 +1,315 @@
+func (p retentionPolicy) RetentionPeriod() RetentionPeriod { return p.p }
+func (p retentionPolicy) Equal(other RetentionPolicy) bool {
+	return p.r.Equal(other.Resolution()) && p.p.Equal(other.RetentionPeriod())
+}
+
+type retentionRule struct {
+	policies                []RetentionPolicy
+	cutoffTime, cutoverTime time.Time
+}
+
+func (r *retentionRule) RetentionPolicies() []RetentionPolicy { return r.policies }
+func (r *retentionRule) CutoverTime() time.Time               { return r.cutoverTime }
+func (r *retentionRule) CutoffTime() time.Time                { return r.cutoffTime }
+
+func (r *retentionRule) SetRetentionPolicies(p []RetentionPolicy) RetentionRule {
+	r.policies = p
+	return r
+}
+
+func (r *retentionRule) SetCutoverTime(t time.Time) RetentionRule {
+	r.cutoverTime = t
+	return r
+}
+
+func (r *retentionRule) SetCutoffTime(t time.Time) RetentionRule {
+	r.cutoffTime = t
+	return r
+}
+
+type query struct {
+	xtime.Range
+	RetentionPolicy
+}
+
+type retentionQueryPlanner struct {
+	log   xlog.Logger
+	clock clock.Clock
+}
+
+func (p retentionQueryPlanner) buildRetentionQueryPlan(from, until time.Time, rules []RetentionRule) ([]query, error) {
+	now := p.clock.Now()
+
+	p.log.Debugf("building query plan from %v to %v over %d rules", from, until, len(rules))
+
+	var queries []query
+
+	// Assumes rules are sorted by cutoff time, with most recent cutoff time first.
+	for _, r := range rules {
+		if !r.CutoffTime().IsZero() && r.CutoffTime().Before(from) {
+			// We've reached a rule that end before the start of the query - no more rules will apply
+			p.log.Debugf("rule cutoff time %s before query start %s, stopping build", r.CutoffTime(), from)
+			break
+		}
+
+		if !r.CutoverTime().IsZero() && r.CutoverTime().After(until) {
+			// We've reached a rule that doesn't apply by the time the query ends - keep going until we find
+			// an older rule which might apply
+			p.log.Debugf("rule cutover time %s after query end %s, skipping rule", r.CutoverTime(), until)
+			continue
+		}
+
+		// Bound the application of the rule by the query range
+		ruleStart, ruleEnd := from, until
+		if !r.CutoffTime().IsZero() && r.CutoffTime().Before(ruleEnd) {
+			ruleEnd = r.CutoffTime()
+		}
+
+		if !r.CutoverTime().IsZero() && r.CutoverTime().After(ruleStart) {
+			ruleStart = r.CutoverTime()
+		}
+
+		p.log.Debugf("rule applies from %s until %s", ruleStart, ruleEnd)
+
+		// Determine queries based on the retention policies in the rule.  Assumes
+		// that the retention policies are ordered by retention period, shortest
+		// retention period first.  Stops as soon as we've reached the last policy
+		// whose retention period falls within the rule or query bounds
+		lastPolicyStart := ruleEnd
+		policies := r.RetentionPolicies()
+		for n, policy := range policies {
+			query := query{
+				RetentionPolicy: policy,
+				Range: xtime.Range{
+					End:   lastPolicyStart,
+					Start: now.Add(-policy.RetentionPeriod().Duration()),
```

Comment:
Is there any possibility that `lastPolicyStart <= now.Add(-policy.RetentionPeriod().Duration())` here?  i.e. end <= start?